### PR TITLE
Add support for redirect filters

### PR DESCRIPTION
**Pull request by AliceProxy**
Adds support for gateway API Redirect Filters.
This is my first PR to the repo so feedback is greatly appreciated .

Fixes: https://github.com/envoyproxy/gateway/issues/159

### CODE REVIEW

**Comment by danehans** [comment]
@AliceProxy please rebase when you have a moment.

**Comment by AliceProxy** [comment]
Sorry about the recent pushes. I thought I was done with this but realized there were some problems that I needed to fix and I added some better test coverage.

**Comment by danehans** [comment]
@skriss do you mind reviewing since you have a good understanding of this code?

**Comment by danehans** [comment]
@AliceProxy can you investigate the CI failure?

**Comment by skriss** [comment]
Yep it's on my list to review, just need to find some time!

**Comment by AliceProxy** [comment]
> @AliceProxy can you investigate the CI failure?

@danehans The most recent commit should resolve the CI failure.

File: internal/ir/xds.go
Comment (arkodg) [inline internal/ir/xds.go:134]: can you update the order of the fields to reflect the priority mentioned in the comment

File: internal/ir/xds.go
Comment (arkodg) [inline internal/ir/xds.go:183]: can you add `Validate` methods for these new structs ?

File: internal/xds/translator/route.go
Comment (arkodg) [inline internal/xds/translator/route.go:11]: can you also add some tests here https://github.com/envoyproxy/gateway/blob/791408311703b662f56f26fcadf2b2e77154f34a/internal/xds/translator/translator_test.go#L26 for redirect and direct response ?

File: internal/gatewayapi/translator.go
Comment (skriss) [inline internal/gatewayapi/translator.go:559]: nit, there are constants for the filter types defined in Gateway API:

```suggestion
						case v1beta1.HTTPRouteFilterRequestRedirect:
```

File: internal/gatewayapi/translator.go
Comment (skriss) [inline internal/gatewayapi/translator.go:552]: It might make sense to move this out of the loop on `rule.Matches` since the same set of filters apply for all matches in a given Gateway API route rule, that way we can avoid processing the filters multiple times, once per match. But, not a big deal right now and we can always revisit.

File: internal/gatewayapi/translator.go
Comment (skriss) [inline internal/gatewayapi/translator.go:581]: I see that the [Gateway API docs](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRequestRedirectFilter) reference setting an "Attached" condition in many of these error cases with a reason "UnsupportedValue", but that condition type doesn't exist for HTTPRoutes. I'm guessing that should actually be the "Accepted" condition, which is listed as having "UnsupportedValue" as a reason. (ref. https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.RouteConditionType).

We should open an issue/contribute a fix to Gateway API for those docs, and assuming that goes through, update these conditions to use "Accepted".

File: internal/gatewayapi/translator.go
Comment (skriss) [inline internal/gatewayapi/translator.go:904]: Contour's using https://github.com/projectcontour/contour/blob/e067e1c59b44cd7b859b6ca8b958bd0018edc789/internal/gatewayapi/listeners.go#L161-L179 to do the valid hostname check (plus the same IP address check you have here) - if that upstream function works, we could use that in place of the custom validation code here.

File: internal/gatewayapi/translator.go
Comment (LukeShu) [inline internal/gatewayapi/translator.go:903]: Any reason not to use [`k8s.io/apimachinery/pkg/util/validation.IsDNS1123Subdomain`](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/validation#IsDNS1123Subdomain)?

File: internal/gatewayapi/translator.go
Comment (LukeShu) [inline internal/gatewayapi/translator.go:920]: My gut reaction is that regexes are big and slow, and `strings.Contains("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.-", char)` would be comparatively faster.

File: internal/gatewayapi/translator.go
Comment (skriss) [inline internal/gatewayapi/translator.go:607]: nit, there are constants for the path type options upstream as well (https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/httproute_types.go#L759-L775)

File: internal/gatewayapi/translator.go
Comment (LukeShu) [inline internal/gatewayapi/translator.go:923]: Any reason not to use `for i, char := range hostname {`?

File: internal/gatewayapi/translator.go
Comment (LukeShu) [inline internal/gatewayapi/translator.go:961]: Prefer to use `%q` to handle quoting for you, instead of `'%s'`.

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks @AliceProxy, overall this looks great! Had a few initial comments, still looking through the test cases.

File: internal/gatewayapi/translator.go
Comment (LukeShu) [inline internal/gatewayapi/translator.go:957]: So to sorta answer my earlier question: It looks like Kube's `IsDNS1123Subdomain` doesn't check the lengths of individual labels.  We should probably send kubenetes/kubernetes.git a PR to fix that.

But! It looks like yours doesn't quite to it right either.  If I'm reading this right, it'd allow the _last_ label to be longer than 63 chars.

File: internal/ir/xds.go
Comment (LukeShu) [inline internal/ir/xds.go:200]: Nit: These are complete sentences, so they should have punctuation. (I wouldn't bother with this comment on a plain comment in a function or something, but these are doc comments.)

**Comment by skriss** [comment]
One other note, just for future reference - our initial goal for EG is to meet "core" Gateway API conformance. As such, I think it'd be fine to start by only implementing the parts of these filters that are required for core conformance, and save "extended" features for later (that's the approach that's been taken with most of the existing Gateway API code). It might help keep the PRs a bit smaller.

File: internal/gatewayapi/translator.go
Comment (LukeShu) [inline internal/gatewayapi/translator.go:650]: We're ending up with a lot of silly code that looks like this. Perhaps we should add a utility function

```go
func ptrTo[T any](x T) *T {
    return &x
}
```

File: internal/gatewayapi/translator.go
Comment (AliceProxy) [reply internal/gatewayapi/translator.go:957]: I'll fix the issue in my code and then also open a PR into kubernetes/kubetnetes.git, but for now I'll just leave a TODO comment saying that we should switch to `IsDNS1123Subdomain` once that submission lands.

File: internal/gatewayapi/translator.go
Comment (LukeShu) [reply internal/gatewayapi/translator.go:903]: I sorta hate that this returns a `string` instead of an `error`.

Comment (AliceProxy) [reply internal/gatewayapi/translator.go:903]: See my response to your other comment re: using `IsDNS1123Subdomain`

File: internal/gatewayapi/translator_test.go
Comment (LukeShu) [inline internal/gatewayapi/translator_test.go:184]: At first I thought "that doesn't work, does it!? You need to use `assert.EqualError`, don't you?", then I noticed that `err` is actually a string.

**Comment by arkodg** [comment]
@AliceProxy thanks for adding the tests in the IR and xDS libs ! LGTM from my end wrt code in those libs, hoping others can review the `gatewayapi/translator` diffs

File: internal/gatewayapi/translator.go
Comment (AliceProxy) [reply internal/gatewayapi/translator.go:923]: Probably just had a brainfart thinking about indexes too much

File: internal/gatewayapi/translator.go
Comment (AliceProxy) [reply internal/gatewayapi/translator.go:904]: Re Luke's comments about `IsDNS1123Subdomain` since it does not check label length I think PR a fix for that in `IsDNS1123Subdomain` and then switch to using that when the change is accepted would be appropriate here :thinking:

File: internal/gatewayapi/translator.go
Comment (arkodg) [reply internal/gatewayapi/translator.go:957]: another option to consider https://github.com/istio/istio/blob/da6178604559bdf2c707a57f452d16bee0de90c8/pkg/config/labels/instance.go#L95

File: internal/gatewayapi/translator.go
Comment (skriss) [reply internal/gatewayapi/translator.go:581]: I submitted https://github.com/kubernetes-sigs/gateway-api/pull/1382 for this

File: internal/gatewayapi/translator.go
Comment (AliceProxy) [reply internal/gatewayapi/translator.go]: Actually I've switched to using `IsDNS1123Subdomain` along with the check for label length (and a test case for the last label as well)

File: internal/gatewayapi/translator.go
Comment (AliceProxy) [reply internal/gatewayapi/translator.go]: Switched to `IsDNS1123Subdomain` which covers most of these checks.

File: internal/gatewayapi/translator.go
Comment (AliceProxy) [reply internal/gatewayapi/translator.go]: I've actually switched to using `IsDNS1123Subdomain` along with Arko's suggestion to include a check for individual label length.

File: internal/gatewayapi/translator.go
Comment (AliceProxy) [reply internal/gatewayapi/translator.go:933]: We don't need the wildcard check since that's invalid for a redirect, but the logic is now a combination of `IsDNS1123Subdomain` and a check on the individual label length.

File: internal/gatewayapi/translator.go
Comment (AliceProxy) [reply internal/gatewayapi/translator.go]: Very true, but I kind of feel like this should be a separate PR after this just to keep the logical size smaller.

File: internal/gatewayapi/translator_test.go
Comment (AliceProxy) [reply internal/gatewayapi/translator_test.go]: It's an `error` now :slightly_smiling_face:

File: internal/gatewayapi/translator.go
Comment (AliceProxy) [reply internal/gatewayapi/translator.go]: Do you want me to change them all to `Accepted` in this PR? I'd be curious to see what the gateway API folks have to say about it first :thinking: 

I'm also wondering if `Accepted` with a status of `False` is the right move on this since the `HTTPRoute` itself would actually be accepted, just the filter may be invalid/ignored. Happy to change them in this PR if you anticipate the Gateway API folks agreeing.

Comment (skriss) [reply internal/gatewayapi/translator.go]: Let's keep as-is for now, I'll create an issue to track the follow-up here, once we see what the Gateway API folks have to say about it. Ultimately we should do whatever is defined in the spec / the conformance tests, I think in this case there was just a typo.

Comment (skriss) [reply internal/gatewayapi/translator.go]: https://github.com/envoyproxy/gateway/issues/357

File: internal/gatewayapi/testdata/httproute-with-redirect-filter-invalid-status.out.yaml
Comment (skriss) [inline internal/gatewayapi/testdata/httproute-with-redirect-filter-invalid-status.out.yaml:85]: Yeah, I think the expected behavior here is unclear in the Gateway API spec -- we can drive some clarity upstream, but doesn't need to block this PR now.

**Review by skriss — APPROVED** [review]
LGTM! Will leave for other reviewers to merge when ready.

**Pull request merged**
